API-Samples/init_texture/init_texture.cpp

```diff
@@ -111,120 +114,115 @@ int sample_main(int argc, char *argv[]) {
     assert(res == VK_SUCCESS);
+
     /* Read the ppm file into the mappable image's memory */
-    if (!read_ppm(filename.c_str(), texObj.tex_width, texObj.tex_height, layout.rowPitch, (unsigned char *)data)) {
+    if (!read_ppm(filename.c_str(), texObj.tex_width, texObj.tex_height,
+                  texObj.needs_staging ? (texObj.tex_width * 4) : layout.rowPitch, (unsigned char *)data)) {
         std::cout << "Could not load texture file lunarg.ppm\n";
         exit(-1);
     }
 
-    vkUnmapMemory(info.device, mappableMemory);
+    vkUnmapMemory(info.device, texObj.needs_staging ? texObj.buffer_memory : texObj.image_memory);
 
-    if (!needStaging) {
+    if (!texObj.needs_staging) {
         /* If we can use the linear tiled image as a texture, just do it */
-        texObj.image = mappableImage;
-        texObj.mem = mappableMemory;
         texObj.imageLayout = VK_IMAGE_LAYOUT_SHADER_READ_ONLY_OPTIMAL;
         set_image_layout(info, texObj.image, VK_IMAGE_ASPECT_COLOR_BIT, VK_IMAGE_LAYOUT_PREINITIALIZED, texObj.imageLayout,
                          VK_PIPELINE_STAGE_HOST_BIT, VK_PIPELINE_STAGE_FRAGMENT_SHADER_BIT);
     } else {
-        /* The mappable image cannot be our texture, so create an optimally
-         * tiled image and blit to it */
-        image_create_info.tiling = VK_IMAGE_TILING_OPTIMAL;
-        image_create_info.usage = VK_IMAGE_USAGE_TRANSFER_DST_BIT | VK_IMAGE_USAGE_SAMPLED_BIT;
-        image_create_info.initialLayout = VK_IMAGE_LAYOUT_UNDEFINED;
-
-        res = vkCreateImage(info.device, &image_create_info, NULL, &texObj.image);
-        assert(res == VK_SUCCESS);
-
-        vkGetImageMemoryRequirements(info.device, texObj.image, &mem_reqs);
-
-        mem_alloc.allocationSize = mem_reqs.size;
-
-        /* Find memory type - don't specify any mapping requirements */
-        pass = memory_type_from_properties(info, mem_reqs.memoryTypeBits, 0, &mem_alloc.memoryTypeIndex);
-        assert(pass);
-
-        /* allocate memory */
-        res = vkAllocateMemory(info.device, &mem_alloc, NULL, &texObj.mem);
-        assert(res == VK_SUCCESS);
-
-        /* bind memory */
-        res = vkBindImageMemory(info.device, texObj.image, texObj.mem, 0);
-        assert(res == VK_SUCCESS);
-
-        /* Since we're going to blit from the mappable image, set its layout to
-         * SOURCE_OPTIMAL */
-        /* Side effect is that this will create info.cmd */
-        set_image_layout(info, mappableImage, VK_IMAGE_ASPECT_COLOR_BIT, VK_IMAGE_LAYOUT_PREINITIALIZED,
-                         VK_IMAGE_LAYOUT_TRANSFER_SRC_OPTIMAL, VK_PIPELINE_STAGE_HOST_BIT, VK_PIPELINE_STAGE_TRANSFER_BIT);
         /* Since we're going to blit to the texture image, set its layout to
          * DESTINATION_OPTIMAL */
         set_image_layout(info, texObj.image, VK_IMAGE_ASPECT_COLOR_BIT, VK_IMAGE_LAYOUT_UNDEFINED,
```

Comment:
Seems like the initialLayout above should be UNDEFINED in this case. (it would be clearer, if not more correct)